PBFT Documentation

bridger-herman opened the pull request:
Changes introduced:

+ Add Rust documentation for the sawtooth-pbft crate
+ Add overall PBFT description, including:
  + Normal case algorithm operation
  + Special case operations like checkpoints and view changes
  + Installing and running PBFT
  + Future work
+ Update README

agunde406 commented on docs/source/_templates/indexcontent.html:
```diff
@@ -0,0 +1,43 @@
+{#
+ # Licensed under Creative Commons Attribution 4.0 International License
```
This file can probably be removed or at least need to be updated for pbft

agunde406 commented on docs/source/conf.py:
```diff
@@ -0,0 +1,332 @@
+# List of patterns, relative to source directory, that match files and
+# directories to ignore when looking for source files.
+exclude_patterns = [
+    'community/how_to_report_issues.rst',
```
This is core specfic

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+********************
+Introduction to PBFT
+********************
+
```
Add a description of the Sawtooth PBFT algorithm (adapt the RFC's Summary & Motivation content): 


This guide describes a practical Byzantine fault tolerant (PBFT) consensus algorithm for Hyperledger Sawtooth. 

The algorithm uses the consensus API described in a Sawtooth RFC.

The Sawtooth PBFT algorithm is a voting-based consensus mechanism with Byzantine fault tolerance. This algorithm ensures safety and liveness of a network, provided at most floor((n - 1)/3) nodes are faulty, where n is the total number of nodes in the network [1]. The PBFT algorithm is also inherently crash fault tolerant. Another advantage of PBFT is that blocks committed by nodes are final, so there are no forks in the network. This is verified by a "Consensus Seal," which is appended to blocks when they are finalized and checked upon receipt of a new block.

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+********************
+Introduction to PBFT
```
Change to "Introduction to Sawtooth PBFT"

I recommend always using the term "Sawtooth PBFT" to refer to this implementation of the Practical Byzantine Fault Tolerance algorithm.  The term "PBFT" alone should refer only to the general algorithm (as described in Castro and Liskov's paper, etc.).

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+Introduction to PBFT
+********************
+
+`Practical Byzantine Fault Tolerance
```
Add a new subheader above the PBFT details:

About PBFT
=========

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+   result.
+
+
+PBFT and the Blockchain
```
Change title to "Consensus API"

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+how the actual blockchain algorithm works.
+
+
+Terminology
```
I suggest moving the Terminology section to the end of the document. 

You could call it "Glossary" instead (that's what the Sawtooth core docs use), but that's a minor detail.

jsmitchell commented on docs/source/conf.py:
```diff
@@ -0,0 +1,329 @@
+#  dir menu entry, description, category)
+texinfo_documents = [
+    (master_doc, 'SawtoothPBFT', u'Sawtooth PBFT Documentation',
+     author, 'SawtoothPBFT', 'One line description of project.',
```
Maybe replace "One line description of project" with one. :)

jsmitchell commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+  public keys refer to the validator’s public key, usually located in
+  ``/etc/sawtooth/keys/validator.pub``
+
+- ``sawtooth.consensus.pbft.block_duration=200``: How often to try to publish
```
Are all these times in milliseconds?

jsmitchell commented on src/config.rs:
```diff
@@ -59,6 +66,20 @@ impl PbftConfig {
+///
+/// Configuration loads the following settings:
+/// + `sawtooth.consensus.pbft.peers` (required)
+/// + `sawtooth.consensus.pbft.block_duration` (optional, default 200 ms)
```
replicating this units info in the doc would be good

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+|                   | blocks given to it by the Consensus API, and     |
+|                   | starting the consensus process.                  |
++-------------------+--------------------------------------------------+
+| Secondary         | Auxiliary node used for consensus.               |
```
I suggest changing "node" to "nodes" (so it's clear that there are multiple secondary nodes in a network with 3 or more nodes).

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
++-------------------+--------------------------------------------------+
+| Consensus seal    | Proof that a block underwent consensus           |
++-------------------+--------------------------------------------------+
+| Block duration    | How many seconds to wait in between the creation |
```
Change "the creation of each block" to "publishing each block".

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+| Checkpoint        | Point in time when logs can get garbage          |
+|                   | collected.                                       |
++-------------------+--------------------------------------------------+
+| Checkpoint period | How many client requests in between each         |
```
Change "client requests" to "blocks are published" (or something like that).

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+|                   | sends them through the Consensus API to the      |
+|                   | consensus algorithm.                             |
++-------------------+--------------------------------------------------+
+| Checkpoint        | Point in time when logs can get garbage          |
```
I suggest changing the definition to "Mode when log messages are pruned (garbage collected). Checkpoint mode is triggered when a certain number of blocks have been committed." (or published?)

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+| High water mark   | Low water mark plus the desired maximum size of  |
+|                   | nodes’ message logs.                             |
++-------------------+--------------------------------------------------+
+| View              | The scope of PBFT when the current primary is in |
```
I suggest changing "scope of PBFT" to "period".

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
++-------------------+--------------------------------------------------+
+| :math:`f`         | The maximum number of faulty nodes.              |
++-------------------+--------------------------------------------------+
+| :math:`v`         | The current view number.                         |
```
I suggest adding " (how many primary nodes changes have occurred)" or something similar.

annechenette commented on docs/source/message-types.rst:
```diff
@@ -0,0 +1,67 @@
+*************
+Message Types
```
Suggestion: Change to "Consensus-related Messages"

(The message types are described/listed in a different section.)

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+correct number of them) need to be handled by the PBFT consensus engine.
+
+
+On-Chain Settings
```
This is very helpful information - thanks for including it!

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+family <https://sawtooth.hyperledger.org/docs/core/releases/latest/transaction_family_specifications/settings_transaction_family.html>`__,
+shown with their respective defaults:
+
+- ``sawtooth.consensus.pbft.peers`` (required): JSON-formatted string of
```
I recommend adding a definition before the format info:

     - ``sawtooth.consensus.pbft.peers`` (required): List of nodes in the network, as a JSON-formatted string of ...

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+
+- ``sawtooth.consensus.pbft.peers`` (required): JSON-formatted string of
+  ``[<public-key-1>, <public-key-2>, ..., <public-key-n>]`` mappings. For
+  instance, ``['\\\"'$$(cat
```
I suggest putting the example on a separate line for legibility, using a ``..code-block`` directive to get fixed-width font. 


     This string defines a two-node network:

     .. code-block: none

       ``['\\\"'$$(cat /etc/sawtooth/keys/validator-0.pub)'\\\"','\\\"'$$(cat  /etc/sawtooth/keys/validator-1.pub)'\\\"']``

Be careful with the spacing - indent these line so that the example is "in" the first bullet.

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+     own
+
+
+Normal Case Operation
```
I suggest removing this heading, and bumping up the level of the following headers (Message Types and States).  Normal Operation (aka Normal Mode) is described later.

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+=====================
+
+Explanation of Message Types
+----------------------------
```
Change to "Message Types" and underline with ====s instead of ----s.  (This section is a peer of "Predicates").

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+  faulty.
+
+States
+------
```
Underline with ====s instead of ----s.  (This section is a peer of "Predicates").

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+  current working block.
+
+The states, state transitions, and actions that the algorithm takes are
+represented in the following diagram:
```
Add info on the blue ovals and gray boxes:

... represented in the following diagram. States in normal mode are shown in blue ovals. Actions are shown in gray rectangles.

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+The states, state transitions, and actions that the algorithm takes are
+represented in the following diagram:
+
+.. figure:: images/pbft_states.png
```
In the diagram, I suggest dropping the "ing" from Checkpointing and ViewChanging to consistent with the terms in the text: Checkpoint (mode) and View Change (mode).

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+.. figure:: images/pbft_states.png
+   :alt: PBFT states
+
+   Possible states in the PBFT algorithm. The outer ring represents normal
```
I would delete these two sentences, because this info is already stated above.  If they stay, change "normal case operation" to "normal mode".

annechenette commented on docs/source/introduction.rst:
```diff
@@ -0,0 +1,132 @@
+The Consensus API abstracts out interactions with the validator, and only
+provides the consensus engine information necessary to the consensus process.
+
+See `Normal Case Operation <operation.html#normal-case-operation>`__ for more details on
```
This link will need to match the new title and location (at line 214), if you accept my suggestions below (lines 115 and 214).

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+- Establish timers and counters for checkpoint periods and block durations,
+  which are loaded from the on-chain settings
+
+Event Loop
```
Change to "Normal Mode" (because this is where the normal-mode operation is described).

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+    - Sequence number is within the sequential bounds of the log (low and high
+      water marks)
+
+    Once the ``PrePrepare`` is accepted:
```
This should be item 3 (not nested under 2).  

Tip: In theory, you don't need to hard-code the item numbers.  Instead, start the line with #. to continue the list numbering.

You can start using #. at item 2.

annechenette commented on docs/source/operation.rst:
```diff
@@ -0,0 +1,276 @@
+    If the ``PrePrepare`` is determined to be invalid, then start a view
+    change.
+
+3. Receive ``Prepare`` messages, and check them all against their associated
```
Change 3. to #. so that you get autonumbering (here and all following numbered items).

annechenette commented on docs/source/special-operations.rst:
```diff
@@ -0,0 +1,46 @@
+******************
+Special Operations
```
I would remove the title "Special Operations".  The two sections here should be at the same level as "Normal Mode" (or "Normal Operation").  

From a formatting perspective, it might make sense to move this file's contents into ``operation.rst``.

annechenette commented on docs/source/special-operations.rst:
```diff
@@ -0,0 +1,46 @@
+Special Operations
+******************
+
+Garbage Collection
```
Change to "Checkpoint Mode", because that's what is shown in the diagram in operations.rst.

The title here should be consistent with the one used in ``operation.rst`` - that is, if it's "Normal Mode" there, this one should be "Checkpoint Mode".  If you use "Normal Operation", this should be "Checkpoint Operation".

annechenette commented on docs/source/special-operations.rst:
```diff
@@ -0,0 +1,46 @@
+
+2. Once a server receives :math:`2f + 1` ``ViewChange`` messages (including
+   its own), it changes its own view to :math:`v + 1`, and resumes ``Normal``
+   operation. The new primary node’s ID is :math:`p = v \mod n`.
```
I suggest adding a sentence on how the next primary node is chosen.  It wasn't clear to me (until I talked to you) that the nodes go in order: Node 0 is the first primary, then node 1, then node 2, and so on.

annechenette commented on docs/source/installing-and-running-pbft.rst:
```diff
@@ -0,0 +1,57 @@
+Installing and Running PBFT
+***************************
+
+This short guide assumes that you have `Docker and Docker Compose
```
Change to 2 paragraphs:

     Prerequisites: `Docker and Docker Compose <https://www.docker.com/>`__ must be installed.

     Use the following steps to reproduce ....

annechenette commented on docs/source/installing-and-running-pbft.rst:
```diff
@@ -0,0 +1,57 @@
+<https://www.docker.com/>`__ installed. Use the following steps to reproduce
+the liveness test on a four node network:
+
+-  Clone the PBFT repo: ``git clone https://github.com/bitwiseio/sawtooth-pbft.git``
```
Use a numbered list (not bullets).  The first item starts with ``1.`` then the rest can start with ``#.``

annechenette commented on docs/source/installing-and-running-pbft.rst:
```diff
@@ -0,0 +1,57 @@
+   # You can optionally use this to speed up your build times, if you have a cargo-registry Docker volume set up:
+   # docker run -v $(pwd):/project/sawtooth-pbft -v cargo-registry:/root/.cargo/registry -it sawtooth-dev-pbft bash
+
+-  Now, inside the ``sawtooth-dev-pbft`` interactive shell, do
```
I would change this to:

     Connect to the ``sawtooth-dev-pbft`` interactive shell, then run:

annechenette commented on docs/source/installing-and-running-pbft.rst:
```diff
@@ -0,0 +1,57 @@
+
+   tests/pbft.sh
+
+This first builds a few docker images, then start up a network of four nodes,
```
Change "This" to "This script"

Also fix the verb tense:
- Change "then start up" to "then starts up"
- Change ", and go through" to " and goes through" (note that there shouldn't be a comma before "and").

bridger-herman replied:
Ah yes, my habitual comma mistakes. Had to show up somewhere here...

annechenette commented on docs/source/installing-and-running-pbft.rst:
```diff
@@ -0,0 +1,57 @@
+
+This short guide assumes that you have `Docker and Docker Compose
+<https://www.docker.com/>`__ installed. Use the following steps to reproduce
+the liveness test on a four node network:
```
Change "reproduce the liveness test on a four node network" to:

     start a four-node network and run a liveness test

annechenette commented on docs/source/installing-and-running-pbft.rst:
```diff
@@ -0,0 +1,57 @@
+
+   cargo build
+
+-  Once the project finishes building, exit the Docker shell and run
```
I'd drop the "Docker" and just say "exit the shell".  (That would avoid the inconstancy of saying "Docker shell" here but "interactive shell" in the previous step.)

annechenette commented on docs/source/installing-and-running-pbft.rst:
```diff
@@ -0,0 +1,57 @@
+
+-  Run the following commands:
+
+::
```
The indenting is off for the commands in this list.  The Sawtooth core uses the ``.. code-block`` directive that's indented at the same level as the text for the list item.

For example: 

     #. Run the following commands:     <- No indent

          .. code-block:: console     <- Indent 3 spaces

              cd sawtooth-pbft        <- Indent 6 spaces
              ...

annechenette commented on docs/source/installing-and-running-pbft.rst:
```diff
@@ -0,0 +1,57 @@
+
+   docker run -v $(pwd):/project/sawtooth-pbft -it sawtooth-dev-pbft bash
+
+   # You can optionally use this to speed up your build times, if you have a cargo-registry Docker volume set up:
```
Delete the # at the start of this line.  Instead, indent the line 3 spaces so it lines up with the content of this (numbered) step.

annechenette commented on docs/source/installing-and-running-pbft.rst:
```diff
@@ -0,0 +1,57 @@
+   docker run -v $(pwd):/project/sawtooth-pbft -it sawtooth-dev-pbft bash
+
+   # You can optionally use this to speed up your build times, if you have a cargo-registry Docker volume set up:
+   # docker run -v $(pwd):/project/sawtooth-pbft -v cargo-registry:/root/.cargo/registry -it sawtooth-dev-pbft bash
```
Format this line inside a ``.. code-block:: console``, like the other commands.  Note that there must be a blank line after ``.. code-block``, and the command must be indented to line up with the "c".

annechenette commented on docs/source/pbft-source-documentation.rst:
```diff
@@ -0,0 +1,38 @@
+*************************
+PBFT Source Documentation
```
Change the title to:

     Sawtooth PBFT API Reference

(This is consistent with the Sawtooth core docs.)

bridger-herman replied:
Since PBFT is a binary application (not a library), do you think it still makes sense to use the term "API" here? I'll put it in for now, and change it if anyone comes up with something better - I can't think of anything off the top of my head.

annechenette commented on docs/source/pbft-source-documentation.rst:
```diff
@@ -0,0 +1,38 @@
+
+    <p class="large-text">
+        <a href="pbft_doc/sawtooth_pbft/index.html">
+            PBFT rustdocs
```
Hmmmm...

I would move this link after the introductory paragraph.  I would also reword the link text to:

     Sawtooth PBFT API (rustdocs)

Or something like that.

annechenette commented on docs/source/pbft-source-documentation.rst:
```diff
@@ -0,0 +1,38 @@
+<https://github.com/ethereum/EIPs/issues/650>`__ from the Ethereum network.
+
+**Dependencies:**
+
```
I suggest an intro sentence here:

     Dependencies are defined _{specified?}_ in ``Cargo.toml``.

Better yet, you could just have this sentence with a link to the file in master, rather than duplicating the contents here. (Because the contents might change.)

If the contents need to stay, there might be a way to auto-include the relevant lines from ``Cargo.toml``.  If you'd like to try this, ask me for more info. (I haven't done this yet, but I think it would work.)

bridger-herman commented:
@jsmitchell - Addressed your comments; and added more info about the consensus seal TODO into the "Future Work" section.
@chenette - Addressed your comments also. FYI, I replied to one of them regarding naming of the Rust documentation, but GitHub is hiding it since I changed those lines.

bridger-herman commented:
@chenette I also added one more commit that adds a diagram about the messages passed during `Normal` mode (similar to the one from the old RFC).

annechenette commented on README.md:
```diff
@@ -1,5 +1,8 @@
 # Sawtooth PBFT
 
+This repository contains an implementation of the Practical Byzantine Fault
+Tolerant (PBFT) consensus algorith for Hyperledger Sawtooth.
```
Typo: algorithm

annechenette commented on docs/source/glossary.rst:
```diff
@@ -0,0 +1,84 @@
++-------------------+--------------------------------------------------+
+| :math:`f`         | The maximum number of faulty nodes.              |
++-------------------+--------------------------------------------------+
+| :math:`v`         | The current view number (How many primary node   |
```
how (lower-case h)

annechenette commented on docs/source/algorithm-operation.rst:
```diff
@@ -0,0 +1,158 @@
+   their view change timers.
+
+#. If ``block_duration`` has elapsed, then try to ``summarize_block()`` with the
+   current working block. If the working block is not ready,
```
Remove the comma after "ready"
